ledger/src/blockstore.rs

```diff
@@ -2664,71 +2681,21 @@ impl Blockstore {
-        if slot > next_max_slot {
-            let mut starting_iterator = self.address_signatures_cf.iter(IteratorMode::From(
-                (starting_primary_index, address, slot, Signature::default()),
+        let mut address_signatures_iter_timer = Measure::start("iter_timer");
+        let mut iterator = self
+            .address_signatures_cf
+            .iter_filtered(IteratorMode::From(
+                (address, slot, 0, Signature::default()),
                 IteratorDirection::Reverse,
             ))?;
 
-            // Iterate through starting_iterator until limit is reached
-            while address_signatures.len() < limit {
-                if let Some(((i, key_address, slot, signature), _)) = starting_iterator.next() {
-                    if slot == next_max_slot || slot < lowest_slot {
-                        break;
-                    }
-                    if i == starting_primary_index
-                        && key_address == address
-                        && slot >= first_available_block
-                    {
-                        if self.is_root(slot) || confirmed_unrooted_slots.contains(&slot) {
-                            address_signatures.push((slot, signature));
-                        }
-                        continue;
-                    }
-                }
-                break;
-            }
-
-            // Handle slots that cross primary indexes
-            if next_max_slot >= lowest_slot {
-                let mut signatures =
-                    self.find_address_signatures_for_slot(address, next_max_slot)?;
-                signatures.reverse();
-                address_signatures.append(&mut signatures);
-            }
-        }
-        starting_primary_index_iter_timer.stop();
-
-        // Iterate through next_iterator until limit is reached
-        let mut next_primary_index_iter_timer = Measure::start("next_primary_index_iter_timer");
-        let mut next_iterator = self.address_signatures_cf.iter(IteratorMode::From(
-            (next_primary_index, address, slot, Signature::default()),
-            IteratorDirection::Reverse,
-        ))?;
+        // Iterate through starting_iterator until limit is reached
         while address_signatures.len() < limit {
-            if let Some(((i, key_address, slot, signature), _)) = next_iterator.next() {
-                // Skip next_max_slot, which is already included
-                if slot == next_max_slot {
-                    continue;
-                }
+            if let Some(((key_address, slot, _transaction_index, signature), _)) = iterator.next() {
                 if slot < lowest_slot {
                     break;
                 }
-                if i == next_primary_index
-                    && key_address == address
-                    && slot >= first_available_block
-                {
+                if key_address == address && slot >= first_available_block {
```

Comment:
Instead of checking both:
- `slot < lowest_slot` and
- `slot >= first_available_block`

I think we can take `lowest_slot.min(first_available_block)` outside of the loop and then only check that. Or, where `lowest_slot` is defined (the match statement on `until`), use `first_available_block` instead of `0`.

Either way, no need to iterate the data where `lowest_slot < slot < first_available_block`; this could be a non-trivial amount of data given how the special columns are cleaned and that this function does not check `lowest_cleanup_slot`.

Lastly, feel free to defer this outside of this PR (maybe we should put together a TODO list of follow-ups 😄 ); just jotting these down while we're spending the time to examine this code so closely

Reply:
Split out to its own PR: https://github.com/solana-labs/solana/pull/33556